skydriver/images.py

```diff
@@ -0,0 +1,111 @@
+"""Utilities for dealing with docker/cvmfs/singularity images."""
+
+import re
+from pathlib import Path
+
+import cachetools.func
+import requests
+
+from .config import LOGGER
+
+# ---------------------------------------------------------------------------------------
+# constants
+
+
+_IMAGE = "skymap_scanner"
+_SKYSCAN_DOCKER_IMAGE_NO_TAG = f"icecube/{_IMAGE}"
+
+DOCKERHUB_API_URL = (
+    f"https://hub.docker.com/v2/repositories/{_SKYSCAN_DOCKER_IMAGE_NO_TAG}/tags"
+)
+
+# cvmfs singularity
+_SKYSCAN_CVMFS_SINGULARITY_IMAGES_DPATH = Path(
+    "/cvmfs/icecube.opensciencegrid.org/containers/realtime/"
+)
+VERSION_REGEX_MAJMINPATCH = re.compile(r"\d+\.\d+\.\d+")
+VERSION_REGEX_PREFIX_V = re.compile(r"v\d+(\.\d+(\.\d+)?)?")
+VERSION_REGEX_MAJ_OR_MAJMIN = re.compile(r"\d+(\.\d+)?")
+
+# clientmanager
+CLIENTMANAGER_IMAGE_WITH_TAG = "ghcr.io/wipacrepo/skydriver:latest"
+
+
+# ---------------------------------------------------------------------------------------
+# getters
+
+
+def get_skyscan_cvmfs_singularity_image(tag: str) -> Path:
+    """Get the singularity image path for 'tag' (assumes it exists)."""
+    return _SKYSCAN_CVMFS_SINGULARITY_IMAGES_DPATH / f"{_IMAGE}:{tag}"
+
+
+def get_skyscan_docker_image(tag: str) -> str:
+    """Get the docker image + tag for 'tag' (assumes it exists)."""
+    return f"{_SKYSCAN_DOCKER_IMAGE_NO_TAG}:{tag}"
+
+
+# ---------------------------------------------------------------------------------------
+# utils
+
+
+@cachetools.func.ttl_cache(ttl=5 * 60)
+def pseudonym_to_full_version_docker_hub(docker_tag: str) -> str:
+    """Get the full-version tag on Docker Hub that has `docker_tag`'s SHA."""
+
+    def _match_find_full_version(digest_sha: str) -> str:
+        # no error handling
+        url = DOCKERHUB_API_URL
+        while True:
+            resp = requests.get(url).json()
+            for img in resp["results"]:
+                if digest_sha != img["digest"]:
+                    continue
+                if VERSION_REGEX_MAJMINPATCH.fullmatch(img["name"]):
```

Comment:
Could I get a comment on (or near) this line that it's excluding non-versioned tags, or something like that?  I had to think a bit to understand what this code is actually doing.

Reply:
It's much more clear now what's going on: [`7567bf9` (#15)](https://github.com/WIPACrepo/SkyDriver/pull/15/commits/7567bf9b94015117f4be36c5bcbea194c4c88f2f)